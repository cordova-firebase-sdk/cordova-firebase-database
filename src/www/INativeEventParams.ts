export interface INativeEventParams {
  args: Array<any>;

  eventType: string;

  listenerId: string;
}
