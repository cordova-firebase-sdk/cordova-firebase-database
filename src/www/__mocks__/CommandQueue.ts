export const execCmd = jest.fn((params: any): Promise<any> => {
  return Promise.resolve(params);
});
