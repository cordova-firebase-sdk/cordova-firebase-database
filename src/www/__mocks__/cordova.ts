export const exec = jest.fn((onSuccess, onError, pluginName, methodName, args): any => {
  setTimeout(onSuccess, 1);
  return {
    args,
    methodName,
    pluginName,
  };
});
