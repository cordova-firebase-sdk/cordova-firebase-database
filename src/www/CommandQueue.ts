import { exec } from "cordova";
// import { Promise } from "es6-promise";
import { nextTick, PluginBase } from "cordova-firebase-core/index";

declare let Promise: any;

/**
 * Parameters for execCmd function
 */
export interface IExecCmdParams {

  /**
   * Plugin's context
   */
  context: PluginBase;

  /**
   * Plugin's name in native side.
   * If omit this, context.id is used.
   */
  pluginName?: string;

  /**
   * Execute method name
   */
  methodName: string;

  /**
   * parameters for native side
   */
  args?: Array<any>;

  /**
   * synchronize options
   */
  execOptions?: {
    sync?: boolean;
    remove?: boolean;
  };

  [key: string]: any;
}

const commandQueue: Array<any> = [];
let isWaitMethod: string;
let isExecuting: boolean;
let executingCnt: number = 0;
const MAX_EXECUTE_CNT: number = 10;

let stopRequested: boolean = false;
window.addEventListener("unload", (): void => {
  stopRequested = true;
}, {
  once: true,
});

export const execCmd = (params: IExecCmdParams): Promise<any> => {
  params.execOptions = params.execOptions || {};
  params.args = params.args || [];
  params.pluginName = params.pluginName || params.context.id;

  // If the instance has been removed, do not execute any methods on it
  // except remove function itself.
  if (params.context.isRemoved && params.execOptions.remove) {
    console.error("[ignore]" + params.context.id + "." + params.methodName + ", because removed.");
    return Promise.resolve();
  }

  // If the overlay is not ready in native side,
  // do not execute any methods except remove on it.
  if (!params.context.isReady && !params.execOptions.remove) {
    console.error("[ignore]" + params.context.id + "." + params.methodName + ", because it's not ready.");
    return Promise.resolve();
  }

  return new Promise((resolve: (...params?: Array<any>) => void, reject: (...params?: Array<any>) => void) => {
    commandQueue.push({
      request: params,

      onSuccess: (...results: Array<any>): void => {
        // -------------------------------
        // success callback
        // -------------------------------

        // Even if the method was successful,
        // but the _stopRequested flag is true,
        // do not execute further code.
        if (!stopRequested) {
          resolve.apply(params.context, results);
        } else {
          // Page will be destroyed.
          return;
        }

        if (params.methodName === isWaitMethod) {
          isWaitMethod = null;
        }

        executingCnt--;
        nextTick(privateExec);
      },

      onError: (...results: Array<any>): void => {
        // -------------------------------
        // error callback
        // -------------------------------
        if (!stopRequested) {
          reject.apply(params.context, results);
        } else {
          // Page will be destroyed.
          return;
        }

        if (params.methodName === isWaitMethod) {
          isWaitMethod = null;
        }

        executingCnt--;
        nextTick(privateExec);
      },

    });

    // In order to execute all methods in safe,
    // the maps plugin limits the number of execution in a moment to 10.
    //
    // Note that Cordova-Android drops has also another internal queue,
    // and the internal queue drops our statement if the app send too much.
    //
    // Also executing too much statements at the same time,
    // it would cause many errors in native side, such as out-of-memory.
    //
    // In order to prevent these troubles, the maps plugin limits the number of execution is 10.
    if (isExecuting || executingCnt >= MAX_EXECUTE_CNT || isWaitMethod || commandQueue.length === 0) {
      return;
    }

    privateExec();
  });
};


const privateExec = (): void => {

  // You probably wonder why there is this code because it's already simular code at the end of the execCmd function.
  //
  // Because the commandQueue might change after the last code of the execCmd.
  // (And yes, it was occurred.)
  // In order to block surely, block the execution again.
  if (isExecuting || executingCnt >= MAX_EXECUTE_CNT || isWaitMethod || commandQueue.length === 0) {
    return;
  }

  isExecuting = true;

  // Execute some execution requests (up to 10) from the commandQueue
  let task: any;
  while (commandQueue.length > 0 && executingCnt < MAX_EXECUTE_CNT) {
    if (!stopRequested) {
      executingCnt++;
    }

    // Pick up the head one.
    task = commandQueue.shift();

    // push out normal tasks if stopRequested becomes true
    if (stopRequested && !task.request.execOptions.remove) {
      executingCnt--;
      continue;
    }

    // Some methods have to block other execution requests.
    if (task.request.execOptions.sync) {
      isWaitMethod = task.request.methodName;
      // console.log(`[sync start] ${task.request.pluginName}.${task.request.methodName}`);
      exec(task.onSuccess, task.onError, task.request.pluginName, task.request.methodName, task.request.args);
      break;
    }

    // console.log(`[start] ${task.request.pluginName}.${task.request.methodName}`);
    exec(task.onSuccess, task.onError, task.request.pluginName, task.request.methodName, task.request.args);
  }

  isExecuting = false;
};
