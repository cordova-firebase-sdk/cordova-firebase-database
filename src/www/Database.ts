import { exec } from "cordova";
import {
  App,
  BaseArrayClass,
  IAppInitializeOptions,
  loadJsPromise,
  PluginBase,
} from "cordova-firebase-core/index";
import { execCmd, IExecCmdParams } from "./CommandQueue";
import { Reference } from "./DataClasses";


declare let Promise: any;
declare let window: any;

export class Database extends PluginBase {

  protected _queue: BaseArrayClass = new BaseArrayClass();

  private _app: App;

  private _options: IAppInitializeOptions;

  private _url: string;

  private _rootRef: Reference;

  constructor(app: App, options: IAppInitializeOptions) {
    super("database");
    if (!app) {
      throw new Error("app must be FirebaseApp instance");
    }
    if (!options) {
      throw new Error("options must be passed");
    }
    if (!options.databaseURL) {
      throw new Error("options.databaseURL must be passed");
    }
    this._app = app;
    this._options = options;
    this._url = options.databaseURL.replace(/(firebaseio.com).*$/, "$1");
    if (this._url.substr(-1) !== "/") {
      this._url += "/";
    }

    this._rootRef = new Reference({
      key: null,
      parent: null,
      pluginName: this.id,
      ref: null,
      url: this._url,
    }, null);
    this._rootRef._privateInit();

    // Bubbling native events
    this._on("nativeEvent", (data: INativeEventParams): void => {
      this._rootRef._trigger.call(this._rootRef, "nativeEvent", data);
    });

    this._queue._on("insert_at", (): void => {
      if (!this._isReady) {
        return;
      }
      if (this._queue._getLength() > 0) {
        const cmd: any = this._queue._removeAt(0, true);
        if (cmd && cmd.context && cmd.methodName) {
          execCmd(cmd).then(cmd.resolve).catch(cmd.reject);
        }
      }
      if (this._queue._getLength() > 0) {
        this._queue._trigger("insert_at");
      }
    });

    // Create one new instance in native side.
    this._one("fireAppReady", (): void => {
      const onSuccess = (result?: any): void => {
        this._isReady = true;
        this._queue._trigger("insert_at");
        this._trigger("ready");
      };
      const onError = (error: any): void => {
        throw new Error(error);
      };
      exec(onSuccess, onError,
      "CordovaFirebaseDatabase", "newInstance",
      [{
        appName: this._app.name,
        id: this.id,
        options: this._options,
      }]);
    });
  }

  public get app(): App {
    return this._app;
  }


  public get url(): string {
    return this._url;
  }


  /**
   * Database.goOffline
   */
  public goOffline(): void {
    this.exec({
      context: this,
      execOptions: {
        sync: true,
      },
      methodName: "goOffline",
    });
  }

  /**
   * Database.goOnline
   */
  public goOnline(): void {
    this.exec({
      context: this,
      execOptions: {
        sync: true,
      },
      methodName: "goOnline",
    });
  }

  /**
   * Database.ref
   * @param [path] - Optional path representing the location the returned Reference will point.
   * If not provided, the returned Reference will point to the root of the Database.
   * @returns Reference
   */
  public ref(path?: string): Reference {

    let key: string = null;
    let url: string = this.url;

    if (typeof path === "string") {

      if (path === "") {
        throw new Error("First argument must not be empty string");
      }

      if (/[\.#$\[\]]/.test(path)) {
        throw new Error("First argument must be a valid firebase URL and the path can't contain \".\", \"#\", \"$\", \"[\", or \"]\".");
      }

      path = path.replace(/\/$/, "");
      key = path.replace(/^.*\//, "") || null;
      if (url.substr(0, 1) !== "/") {
        url += "/" + path;
      }
      url = url.replace(/\/+/g, "/");
      url = url.replace(/https:\//, "https://");
    }

    // Create a reference instance.
    const reference: Reference = new Reference({
      key,
      parent: null,
      pluginName: this.id,
      ref: null,
      url,
    }, this._rootRef);


    this.exec({
      args: [{
        id: reference.id,
        path,
      }],
      context: this,
      methodName: "database_ref",
    }).then((): void => {
      reference._privateInit();
    });

    return reference;
  }


  /**
   * Database.refFromURL
   * https://firebase.google.com/docs/reference/js/firebase.database.Database#refFromURL
   */
  public refFromURL(url: string): Reference {

    if (typeof url !== "string") {
      throw new Error("First argument must be string");
    }

    if (url === "") {
      throw new Error("First argument must not be empty string");
    }

    const tmp: Array<any> = this.url.match(/:\/\/(.+?.firebaseio.com)/);
    if (url.indexOf(tmp[1]) === -1) {
      throw new Error("First argument must be with the current database");
    }
    url = url.replace(/\?.*$/, "");

    const tmpUrl: string = url.replace(/https:\/\/[a-z0-9]+\.firebaseio\.com\//i, "");
    if (/[\.#$\[\]]/.test(tmpUrl)) {
      throw new Error("First argument must be a valid firebase URL and the path can't contain \".\", \"#\", \"$\", \"[\", or \"]\".");
    }


    let path: string = url.replace(/^https:\/\/.+?.firebaseio.com\/?/, "");
    path = path.replace(/\/$/, "");
    path = path.replace(/^\//, "");

    let reference: Reference = null;
    let currentUrl: string = this.url;
    let newRef: Reference = null;
    const refs: Array<Reference> = (path.split(/\//)).map((key: string) => {

      currentUrl += "/" + key;

      newRef = new Reference({
        key,
        parent: reference,
        pluginName: this.id,
        ref: null,
        url: currentUrl,
      }, this._rootRef);

      // Bubbling native events
      this._on("nativeEvent", (...parameters: Array<any>): void => {
        parameters.unshift("nativeEvent");
        newRef._trigger.apply(newRef, parameters);
      });

      this.exec({
        args: [{
          id: newRef.id,
          path,
        }],
        context: this,
        methodName: "database_ref",
      }).then((): void => {
        newRef._privateInit();
      });

      reference = newRef;

      return newRef;
    });
    //
    // this.exec({
    //   args: [{
    //     id: reference.id,
    //     path,
    //   }],
    //   context: this,
    //   methodName: "database_refFromURL",
    // }).then((): void => {
    //   reference._privateInit();
    // });

    return reference;
  }


  private exec(params: IExecCmdParams): Promise<any> {
    return new Promise((resolve: (result: any) => void, reject: (error: any) => void) => {
      params.resolve = resolve;
      params.reject = reject;
      this._queue._push(params);
    });
  }

}



class SecretDatabaseManager {

  public _databases: any = {};

}

if (window.cordova && window.cordova.version) {
  const manager: SecretDatabaseManager = new SecretDatabaseManager();

  Object.defineProperty(window.plugin.firebase, "database", {
    value: (app?: App): Database => {
      if (!app) {
        // Obtains default app
        app = window.plugin.firebase.app();
      }

      // If database is created for the app, returns it.
      // Otherwise, create a new instance.
      let database: Database = manager._databases[app.name];
      if (!database) {
        database = new Database(app, app.options);
        manager._databases[app.name] = database;
      }
      return database;
    },
  });

  Object.defineProperty(window.plugin.firebase.database, "_nativeCallback", {
    enumerable: false,
    value: (dbId: string, listenerId: string, eventType: string, args: Array<any>): void => {

      const dbInstance = window.plugin.firebase.database._DBs[dbId];

      if (dbInstance) {
        dbInstance._trigger("nativeEvent", {
          args,
          eventType,
          listenerId,
        });
      }
    },
  });
}
