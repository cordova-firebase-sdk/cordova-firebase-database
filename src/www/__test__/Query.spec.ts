import { App, LZString } from "cordova-firebase-core/index";
import { Database } from "../Database";
import { DataSnapshot, Query, Reference } from "../DataClasses";
import { INativeEventParams } from "../INativeEventParams";
import { exec } from "../__mocks__/cordova";
import { execCmd } from "../__mocks__/CommandQueue";
const proxy = require("cordova/exec/proxy");

describe("[Query]", () => {

  const commonApp: App = new App("dummyApp", {
    hello: "world",
    databaseURL: "https://dummy.firebaseio.com/"
  });

  const commonDb: Database = new Database(commonApp, commonApp.options);
  commonDb._trigger("fireAppReady");

  beforeEach(() => {
    // Clear all instances and calls to constructor and all methods:
    exec.mockClear();
    execCmd.mockClear();
  });

  describe("constructor", () => {
    it("should reject if params is null", () => {
      expect(() => {
        new Query(null, {
          exec: null,
        });
      }).toThrowErrorMatchingSnapshot();
    });
    it("should reject if params.url is null", () => {
      expect(() => {
        new Query({
          url: null,
          pluginName: null,
        }, null);
      }).toThrowErrorMatchingSnapshot();
    });
    it("should reject if _opts is null", () => {
      expect(() => {
        new Query({
          url: "https://",
          pluginName: "",
        }, null);
      }).toThrowErrorMatchingSnapshot();
    });

  });
  describe("endAt", () => {
    it("should involve native side with correct parameters", (done) => {
      const ref: Reference = commonDb.ref("");
      const value: any = {
        hello: "world",
      };
      const query: Query = ref.endAt(value, "key");
      setTimeout(() => {
        expect(execCmd.mock.calls[0][0].methodName).toBe("database_ref");
        expect(execCmd.mock.calls[0][0].pluginName).toBe(ref.pluginName);
        expect(execCmd.mock.calls[1][0].pluginName).toBe(query.pluginName);
        expect(ref.pluginName).toBe(query.pluginName);
        expect(execCmd.mock.calls[1][0].methodName).toBe("query_endAt");
        expect(execCmd.mock.calls[1][0].args[0].targetId).toBe(ref.id);
        expect(execCmd.mock.calls[1][0].args[0].key).toBe("key");
        expect(execCmd.mock.calls[1][0].args[0].value).toBe(LZString.compressToBase64(JSON.stringify(value)));
        done();
      }, 5);
    });

    it("should fire `listenerId` event if query receive 'nativeEvent'", (done) => {
      const ref: Reference = commonDb.ref("");
      const value: any = {
        hello: "world",
      };
      const query: Query = ref.endAt(value, "key");

      query._one("listenerId", (data: INativeEventParams): void => {
        expect(data).toEqual({
          args: ["0", "1"],
          eventType: "eventType",
          listenerId: "listenerId",
        });
        setTimeout(done, 0);
      });
      ref.root._trigger("nativeEvent", {
        args: ["0", "1"],
        eventType: "eventType",
        listenerId: "listenerId",
      });
    });
  });
  describe("equalTo", () => {
    it("should involve native side with correct parameters", (done) => {
      const ref: Reference = commonDb.ref("");
      const value: any = {
        hello: "world",
      };
      const query: Query = ref.equalTo(value, "key");

      setTimeout(() => {
        expect(execCmd.mock.calls[0][0].methodName).toBe("database_ref");
        expect(execCmd.mock.calls[1][0].methodName).toBe("query_equalTo");
        expect(execCmd.mock.calls[1][0].args[0].key).toEqual("key");
        expect(execCmd.mock.calls[1][0].args[0].queryId).toBe(query.id);
        expect(execCmd.mock.calls[1][0].args[0].targetId).toBe(ref.id);
        expect(execCmd.mock.calls[1][0].args[0].value).toBe(LZString.compressToBase64(JSON.stringify(value)));
        done();
      }, 5);
    });
  });
  describe("isEqual", () => {
    it("should be true", () => {
      const ref1: Reference = commonDb.ref("users");
      const ref2: Reference = commonDb.ref("users");
      expect(ref1.isEqual(ref2)).toBe(true);
    });
  });
  describe("limitToFirst", () => {
    it("should involve native side with correct parameters", (done) => {
      const ref: Reference = commonDb.ref("");
      const value: any = {
        hello: "world",
      };
      const query: Query = ref.limitToFirst(3);
      setTimeout(() => {
        expect(execCmd.mock.calls[0][0].methodName).toBe("database_ref");
        expect(execCmd.mock.calls[1][0].methodName).toBe("query_limitToFirst");
        expect(execCmd.mock.calls[1][0].args[0].limit).toEqual(3);
        expect(execCmd.mock.calls[1][0].args[0].queryId).toBe(query.id);
        expect(execCmd.mock.calls[1][0].args[0].targetId).toBe(ref.id);
        done();
      }, 5);
    });
  });
  describe("limitToLast", () => {
    it("should involve native side with correct parameters", (done) => {
      const ref: Reference = commonDb.ref("");
      const value: any = {
        hello: "world",
      };
      const query: Query = ref.limitToLast(3);
      setTimeout(() => {
        expect(execCmd.mock.calls[0][0].methodName).toBe("database_ref");
        expect(execCmd.mock.calls[1][0].methodName).toBe("query_limitToLast");
        expect(execCmd.mock.calls[1][0].args[0].limit).toEqual(3);
        expect(execCmd.mock.calls[1][0].args[0].queryId).toBe(query.id);
        expect(execCmd.mock.calls[1][0].args[0].targetId).toBe(ref.id);
        done();
      }, 5);
    });
  });

  describe("off()", () => {
    it("should work correctly", (done) => {
      const ref: Reference = commonDb.ref("");
      const value: any = { hello: "world" };
      const query: Query = ref.limitToLast(3);
      const _onSpy = jest.spyOn(query, "_on");
      let triggerred: boolean = false;

      const listener = query.on("value", (snapshot: DataSnapshot): void => {
        triggerred = true;
      });
      query.off("value", listener);

      const dummySnapshot: string = JSON.stringify({
        exists: true,
        exportVal: LZString.compressToBase64(JSON.stringify({ hello: "world" })),
        getPriority: "high",
        key: "key",
        numChildren: 0,
        val: LZString.compressToBase64(JSON.stringify({ hello: "world" })),
      });

      ref.root._trigger("nativeEvent", {
        args: [LZString.compressToBase64(dummySnapshot)],
        eventType: "value",
        listenerId: _onSpy.mock.calls[0][0],
      });
      setTimeout(() => {
        expect(triggerred).toBe(false);
        done();
      }, 5);
    });
    it("should work correctly when context is specified", (done) => {
      const ref: Reference = commonDb.ref("");
      const value: any = { hello: "world" };
      const query: Query = ref.limitToLast(3);
      const _onSpy = jest.spyOn(query, "_on");
      let triggerred: boolean = false;

      const listener = query.on("value", (snapshot: DataSnapshot): void => {
        triggerred = true;
      });
      query.off("value", listener, query);

      const dummySnapshot: string = JSON.stringify({
        exists: true,
        exportVal: LZString.compressToBase64(JSON.stringify({ hello: "world" })),
        getPriority: "high",
        key: "key",
        numChildren: 0,
        val: LZString.compressToBase64(JSON.stringify({ hello: "world" })),
      });

      ref.root._trigger("nativeEvent", {
        args: [LZString.compressToBase64(dummySnapshot)],
        eventType: "value",
        listenerId: _onSpy.mock.calls[0][0],
      });
      setTimeout(() => {
        expect(triggerred).toBe(false);
        done();
      }, 5);
    });
    it("should work correctly even if listener is omitted", (done) => {
      const ref: Reference = commonDb.ref("");
      const value: any = { hello: "world" };
      const query: Query = ref.limitToLast(3);
      const _onSpy = jest.spyOn(query, "_on");
      let triggerred: boolean = false;

      const listener = query.on("value", (snapshot: DataSnapshot): void => {
        triggerred = true;
      });
      query.off("value");

      const dummySnapshot: string = JSON.stringify({
        exists: true,
        exportVal: LZString.compressToBase64(JSON.stringify({ hello: "world" })),
        getPriority: "high",
        key: "key",
        numChildren: 0,
        val: LZString.compressToBase64(JSON.stringify({ hello: "world" })),
      });

      ref.root._trigger("nativeEvent", {
        args: [LZString.compressToBase64(dummySnapshot)],
        eventType: "value",
        listenerId: _onSpy.mock.calls[0][0],
      });
      setTimeout(() => {
        expect(triggerred).toBe(false);
        done();
      }, 5);
    });
    it("should work correctly even if eventType and listener are omitted", (done) => {
      const ref: Reference = commonDb.ref("");
      const value: any = { hello: "world" };
      const query: Query = ref.limitToLast(3);
      const _onSpy = jest.spyOn(query, "_on");
      let triggerred: boolean = false;

      const listener = query.on("value", (snapshot: DataSnapshot): void => {
        triggerred = true;
      });
      query.off();

      const dummySnapshot: string = JSON.stringify({
        exists: true,
        exportVal: LZString.compressToBase64(JSON.stringify({ hello: "world" })),
        getPriority: "high",
        key: "key",
        numChildren: 0,
        val: LZString.compressToBase64(JSON.stringify({ hello: "world" })),
      });

      ref.root._trigger("nativeEvent", {
        args: [LZString.compressToBase64(dummySnapshot)],
        eventType: "value",
        listenerId: _onSpy.mock.calls[0][0],
      });
      setTimeout(() => {
        expect(triggerred).toBe(false);
        done();
      }, 5);
    });
    it("should throw Error if eventType is invalid", () => {
      const ref: Reference = commonDb.ref("");
      const query: Query = ref.limitToLast(3);
      let triggerred: boolean = false;

      const listener = query.on("value", (snapshot: DataSnapshot): void => {
        triggerred = true;
      });
      expect(() => {
        query.off("invalidEvent", listener);
      }).toThrowErrorMatchingSnapshot();

    });
  });
  describe("on()", () => {
    it("should work correctly", (done) => {
      const ref: Reference = commonDb.ref("");
      const value: any = { hello: "world" };
      const query: Query = ref.limitToLast(3);
      const _onSpy = jest.spyOn(query, "_on");

      query.on("value", (snapshot: DataSnapshot): void => {
        expect(snapshot.key).toEqual("key");
        expect(snapshot.getPriority()).toEqual("high");
        done();
      });

      const dummySnapshot: string = JSON.stringify({
        exists: true,
        exportVal: LZString.compressToBase64(JSON.stringify({ hello: "world" })),
        getPriority: "high",
        key: "key",
        numChildren: 0,
        val: LZString.compressToBase64(JSON.stringify({ hello: "world" })),
      });

      ref.root._trigger("nativeEvent", {
        args: [LZString.compressToBase64(dummySnapshot)],
        eventType: "value",
        listenerId: _onSpy.mock.calls[0][0],
      });
    });
    it.only("should work correctly when cancelCallback is provided", (done) => {
      const ref: Reference = commonDb.ref("");
      const value: any = { hello: "world" };
      const query: Query = ref.limitToLast(3);
      const _onSpy = jest.spyOn(query, "_on");
      let triggerred: boolean = false;

      execCmd.mockImplementationOnce((params) => {
        if (params.methodName !== "query_on") {
          return Promise.resolve(params);
        }
        const _nativeCallback = (window as any).plugin.firebase.database._nativeCallback;

        //_nativeCallback(query.id, params.args[0].id);
        console.log(params);

        return Promise.resolve(params);
      });

      query.on("value", (snapshot: DataSnapshot) => {
        triggerred = true;
      }, (error?: any) => {
        expect(error).toThrowErrorMatchingSnapshot();
        done();
      }, query);
    });
  });
  describe("once()", () => {
    it("should work correctly", (done) => {

      const ref: Reference = commonDb.ref("");
      const value: any = { hello: "world" };
      const query: Query = ref.limitToLast(3);
      const _onSpy = jest.spyOn(query, "_one");

      query.once("value", (snapshot: DataSnapshot): void => {
        expect(snapshot.key).toEqual("key1");
        expect(snapshot.getPriority()).toEqual("high");
        setTimeout(done, 0);
      });

      const dummySnapshot1: string = JSON.stringify({
        exists: true,
        exportVal: LZString.compressToBase64(JSON.stringify({ hello: "world" })),
        getPriority: "high",
        key: "key1",
        numChildren: 0,
        val: LZString.compressToBase64(JSON.stringify({ hello: "world" })),
      });

      const dummySnapshot2: string = JSON.stringify({
        exists: true,
        exportVal: LZString.compressToBase64(JSON.stringify({ hello: "world" })),
        getPriority: "normal",
        key: "key2",
        numChildren: 0,
        val: LZString.compressToBase64(JSON.stringify({ hello: "world" })),
      });

      ref.root._trigger("nativeEvent", {
        args: [LZString.compressToBase64(dummySnapshot1)],
        eventType: "value",
        listenerId: _onSpy.mock.calls[0][0],
      });
      ref.root._trigger("nativeEvent", {
        args: [LZString.compressToBase64(dummySnapshot2)],
        eventType: "value",
        listenerId: _onSpy.mock.calls[0][0],
      });
    });
  });
  describe("orderByChild()", () => {
    it("should involve native side with correct parameters", (done) => {
      const ref: Reference = commonDb.ref("");
      const query: Query = ref.orderByChild("children");
      setTimeout(() => {
        expect(execCmd.mock.calls[0][0].methodName).toBe("database_ref");
        expect(execCmd.mock.calls[1][0].methodName).toBe("query_orderByChild");
        expect(execCmd.mock.calls[1][0].args[0].path).toEqual("children");
        expect(execCmd.mock.calls[1][0].args[0].queryId).toBe(query.id);
        expect(execCmd.mock.calls[1][0].args[0].targetId).toBe(ref.id);
        done();
      }, 5);
    });
  });
  describe("orderByKey()", () => {
    it("should involve native side with correct parameters", (done) => {
      const ref: Reference = commonDb.ref("");
      const query: Query = ref.orderByKey();
      setTimeout(() => {
        expect(execCmd.mock.calls[0][0].methodName).toBe("database_ref");
        expect(execCmd.mock.calls[1][0].methodName).toBe("query_orderByKey");
        expect(execCmd.mock.calls[1][0].args[0].queryId).toBe(query.id);
        expect(execCmd.mock.calls[1][0].args[0].targetId).toBe(ref.id);
        done();
      }, 5);
    });
  });
  describe("orderByPriority()", () => {
    it("should involve native side with correct parameters", (done) => {
      const ref: Reference = commonDb.ref("");
      const query: Query = ref.orderByPriority();
      setTimeout(() => {
        expect(execCmd.mock.calls[0][0].methodName).toBe("database_ref");
        expect(execCmd.mock.calls[1][0].methodName).toBe("query_orderByPriority");
        expect(execCmd.mock.calls[1][0].args[0].queryId).toBe(query.id);
        expect(execCmd.mock.calls[1][0].args[0].targetId).toBe(ref.id);
        done();
      }, 5);
    });
  });
  describe("orderByValue()", () => {
    it("should involve native side with correct parameters", (done) => {
      const ref: Reference = commonDb.ref("");
      const query: Query = ref.orderByValue();
      setTimeout(() => {
        expect(execCmd.mock.calls[0][0].methodName).toBe("database_ref");
        expect(execCmd.mock.calls[1][0].methodName).toBe("query_orderByValue");
        expect(execCmd.mock.calls[1][0].args[0].queryId).toBe(query.id);
        expect(execCmd.mock.calls[1][0].args[0].targetId).toBe(ref.id);
        done();
      }, 5);
    });
  });
  describe("startAt()", () => {
    it("should involve native side with correct parameters", (done) => {
      const ref: Reference = commonDb.ref("");
      const query: Query = ref.startAt(3);
      setTimeout(() => {
        expect(execCmd.mock.calls[0][0].methodName).toBe("database_ref");
        expect(execCmd.mock.calls[1][0].methodName).toBe("query_startAt");
        expect(execCmd.mock.calls[1][0].args[0].queryId).toBe(query.id);
        expect(execCmd.mock.calls[1][0].args[0].targetId).toBe(ref.id);
        expect(execCmd.mock.calls[1][0].args[0].value).toBe(LZString.compressToBase64(JSON.stringify(3)));
        done();
      }, 5);
    });
  });
});
