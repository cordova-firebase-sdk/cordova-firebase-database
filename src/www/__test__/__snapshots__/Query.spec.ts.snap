// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Query] constructor should reject if _opts is null 1`] = `"_opts must be provided."`;

exports[`[Query] constructor should reject if params is null 1`] = `"params must be provided."`;

exports[`[Query] constructor should reject if params.url is null 1`] = `"params.url must be provided."`;

exports[`[Query] off() should throw Error if eventType is invalid 1`] = `"eventType must be one of 'value','child_added', 'child_moved', 'child_removed', or 'child_changed'."`;

exports[`[Query] on() should reject error if eventType is not supportted 1`] = `"eventType must be one of 'value','child_added', 'child_moved', 'child_removed', or 'child_changed'."`;

exports[`[Query] on() should work correctly when cancelCallback is provided 1`] = `"received is not a function"`;

exports[`[Query] once() should reject error if eventType is not supportted 1`] = `"eventType must be one of 'value','child_added', 'child_moved', 'child_removed', or 'child_changed'."`;

exports[`[Query] once() should work correctly when failureCallbackOrContext is provided 1`] = `"received is not a function"`;
